enhancements/network/on-prem-dual-stack-vips.md

```diff
@@ -0,0 +1,339 @@
+This should be a relatively simple feature to implement since most of the
+new code, manifests, and templates will be slight modifications of the
+existing loadbalancer ones. The main thing we need to determine is what the
+structure of the install-config and api changes will be. That will have an
+impact on the implementation of these new VIPs, but any of the designs
+proposed below should be doable. It's mostly a question of picking the one
+that will provide the best user experience and, secondarily, the simplest
+implementation.
+
+### Risks and Mitigations
+
+Minimal risk. This is just adding another VIP, something we already have in
+our deployments. The main concern would be logic errors arising out of
+the need to handle multiple VIPs, which (if they happen) will need to be
+addressed as bugs.
+
+## Design Details
+
+This feature will require changes in a few different components:
+
+Installer - New fields will need to be added to install-config for deployers
+to specify the additional VIPs.
+
+API - New fields will need to be added to the platform status section of the
+infrastructure object.
+
+machine-config-operator - The on-prem loadbalancer templates and manifests
+will need to be updated to support the additional VIPs.
+
+baremetal-runtimecfg - There is logic that depends on the ip version of the
+VIP. If there are multiple VIPs, this will have to be modified to account for
+that. In addition, the render code will need the new values wired in.
+
+### install-config
+
+Currently the VIPs are specified as follows:
+
+```
+platform:
+  baremetal:
+    apiVIP: 192.168.111.5
+    ingressVIP: 192.168.111.4
+```
+
+There are a few options for how we could add the new VIPs (we need to pick
+one and move the others to the alternatives section).
+
+#### Just add v6 versions of the VIP fields
+```
+platform:
+  baremetal:
+    apiVIP: 192.168.111.5
+    apiVIPv6: f00::5
+    ingressVIP: 192.168.111.4
+    ingressVIPv6: f00::4
+```
+This is a minimal change, but it leaves us with somewhat of a problem in that
+we can't assume the old fields always contain a v4 address since previous
+releases allowed either v4 or v6. We may be able to deal with that by
+detecting the appropriate ip version for the old fields and populating new
+fields on the api side. E.g. if apiVIP is v4, we populate
+`platformStatus.apiServerInternalIPv4` and if it is v6 we populate
+`apiServerInternalIPv6` instead. That way the only place we need to worry about
+the difference is in the installer. We would need a validation to ensure a v6
+VIP isn't specified in both fields.
+
+Note that this assumes we replace the existing api fields with two new ones
+per VIP and do not reuse the existing ones at all. I think we're going to want
+to do that regardless of what install-config layout we use so we don't need
+logic on the backend in machine-config-operator and baremetal-runtimecfg to
+handle older cluster configs.
+
+#### Deprecate the old fields and add new v4 and v6 fields
```

Comment:
I like this option the best.  I'd say though it's probably good if this is a relatively "quiet" deprecation at least to start; the cost of carrying the auto-detection code seems low.

Reply:
Yeah, I'm looking at this deprecation more as a "don't use this anymore" than "this is going away". From previous discussions with the installer team I'm not sure we're supposed to break old install-configs anyway, so we may have to carry it indefinitely.

Reply:
+1 Agree with Colin. This also kind of sounds like the Platform/PlatformStatus thing we had to do early on. Where old clusters had an empty new field and we had to populate it correctly for born-in 4.1-ish clusters so they wouldnt break. Overall, we can't break people but it's not the first time we had to do something like this so it should be ok.

Reply:
Cool, this is the one I would tend to prefer as well since it matches what I'm proposing for the api side, and it's good to hear we're all in agreement so far.